### DISCUSSION TITLE
Emit cjs with Nitro preset

### DISCUSSION THREAD

**Discussion by BetaHuhn** [nuxt/nuxt#18100] [Questions]
Hey, just gave Nuxt 3 a try and I have to say it's awesome!

I played around a bit with Nitro presets and wanted to create a custom one to emit cjs instead of ESM. Is that possible?

Specifically I am looking to mount the Nuxt app as a express middleware in a commonjs project. I tried to modify the [node preset](https://v3.nuxtjs.org/docs/deployment/presets/node) without any luck. While looking through Nitro's source code I saw the [rollup config](https://github.com/nuxt/framework/blob/main/packages/nitro/src/rollup/config.ts), but I am not proficient enough to modify that to emit cjs instead of esm.

All the existing presets also only emit ESM, the only mention of CJS I found was in this issue https://github.com/nuxt/framework/issues/1895

While I think ESM is awesome, support for it isn't widely available yet and it would be great to have the ability to build the app to CJS to be able to deploy it on more platforms which don't support ESM yet.

Happy to give more context or help anyway I can.

Sending lots of ❤️ to the Nuxt team for their hard work! Thanks!

**Comment by pi0** [comment]
Hi, @BetaHuhn thanks for the kind words! May I ask which platforms you are referring to? Amazon AWS for instance only supports CJS entrypoint but still can use native ESM using dynamic import. It would be possible for nitro-generating CJS but needs deoptimizations like disabling externals tracking.

**Reply by BetaHuhn** [reply]
> May I ask which platforms you are referring to?

For one I want to mount the Nuxt App as an Express middleware inside an existing project which is using CJS. I was also thinking about using Nuxt 3 with [Deta](https://docs.deta.sh/docs/micros/deploy#nodejs) which currently only supports a CJS entrypoint.

> It would be possible for nitro-generating CJS but needs deoptimizations like disabling externals tracking.

Is this configurable at the moment or would this require changes to Nitro?

**Reply by pi0** [reply]
> For one I want to mount the Nuxt App as an Express middleware inside an existing project which is using CJS. I was also thinking about using Nuxt 3 with Deta which currently only supports a CJS entrypoint.

Actually, you can register a lazy middleware for express with dynamic import within CJS context: 

```js
const express = require('express');

const app = express();

app.use('/', (req, res) => {
  import('./.output/server/index.mjs')
    .then((m) => m.default || m)
    .then((handle) => handle(req, res));
});
````

**Reply by BetaHuhn** [reply]
Changed it to:

```js
let nuxt

app.use(async (req, res) => {
	if (!nuxt) {
		nuxt = await import('./.output/server/index.mjs')
	}

	nuxt.handle(req, res)
})
```

which works great in my existing project, thanks!

On Deta this doesn't work though, looks like they don't support ESM at all, even if the entrypoint is CJS.

Is there a way for me to get Nitro to generate CJS directly or would that need changes on your side?
